python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,471 @@
+      TimestampType, TimestamptzType, StringType, BinaryType, UUIDType, FloatType, or DoubleType.
+      num_buckets (int): The number of buckets.
+
+    Raises:
+      ValueError: If a type is provided that is incompatible with a Bucket transform
+    """
+
+    _MAX_32_BITS_INT = 2147483647
+    _INT_TRANSFORMABLE_TYPES = {
+        IntegerType,
+        DateType,
+        LongType,
+        TimeType,
+        TimestampType,
+        TimestamptzType,
+    }
+    _SAME_TRANSFORMABLE_TYPES = {
+        StringType,
+        BinaryType,
+        UUIDType,
+        FloatType,
+        DoubleType,
+    }
+
+    def __init__(self, source_type: Type, num_buckets: int):
+        super().__init__(
+            f"bucket[{num_buckets}]",
+            f"transforms.bucket(source_type={repr(source_type)}, num_buckets={num_buckets})",
+        )
+        self._type = source_type
+        self._num_buckets = num_buckets
+
+        if isinstance(self._type, FixedType) or isinstance(self._type, DecimalType):
+            self._can_transform = lambda t: type(self._type) is type(t)
+        elif self._type in Bucket._SAME_TRANSFORMABLE_TYPES:
+            self._can_transform = lambda t: self._type == t
+        elif self._type in Bucket._INT_TRANSFORMABLE_TYPES:
+            self._can_transform = (
+                lambda t: self._type in Bucket._INT_TRANSFORMABLE_TYPES
+            )
+        else:
+            raise ValueError(f"Cannot bucket by type: {source_type}")
+
+        if (
+            isinstance(self._type, FixedType)
+            or self._type == StringType
+            or self._type == BinaryType
+        ):
+            self._hash_func = lambda v: mmh3.hash(v)
+        elif isinstance(self._type, DecimalType):
+            self._hash_func = lambda v: mmh3.hash(transform_util.decimal_to_bytes(v))
+        elif self._type == FloatType or self._type == DoubleType:
+            # bucketing by Float/Double is not allowed by the spec, but they have hash implementation
+            self._hash_func = lambda v: mmh3.hash(struct.pack("d", v))
+        elif self._type == UUIDType:
+            self._hash_func = lambda v: mmh3.hash(
+                struct.pack(
+                    ">QQ",
+                    (v.int >> 64) & 0xFFFFFFFFFFFFFFFF,
+                    v.int & 0xFFFFFFFFFFFFFFFF,
+                )
+            )
+        else:
+            self._hash_func = lambda v: mmh3.hash(struct.pack("q", v))
+
+    @property
+    def num_buckets(self) -> int:
+        return self._num_buckets
+
+    def apply(self, value) -> Optional[int]:
+        if value is None:
+            return None
+
+        return (self._hash_func(value) & Bucket._MAX_32_BITS_INT) % self._num_buckets
+
+    def can_transform(self, target: Type) -> bool:
+        return self._can_transform(target)
+
+    def result_type(self, source: Type):
+        return IntegerType
+
+
+class TimeTransform(Transform):
```

Comment:
When referring to both Date and Timestamp, it is usually better to use "DateTime" so it is clear that you're not talking about a specific type (date, time or timestamp).

However, this may be moot because I recommend the same thing I said for the bucket transform. Using lambdas doesn't seem to have value to me. I would probably decompose this into more specific classes.

Reply:
SG

Reply:
@rdblue  I tried to split them into 7 classes to avoid lambda but seems quite verbose (each class just override the apply function with 1 line method call). I rewrote this class and improved it to make the logic clear. Can you take another look and see if it is good? Thanks.